internal/components/component.go

```diff
@@ -0,0 +1,127 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package components
+
+import (
+	"encoding/json"
+	"errors"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+var (
+	GrpcProtocol          = "grpc"
+	HttpProtocol          = "http"
+	UnsetPort       int32 = 0
+	PortNotFoundErr       = errors.New("port should not be empty")
+)
+
+type PortRetriever interface {
+	GetPortNum() (int32, error)
+	GetPortNumOrDefault(logr.Logger, int32) int32
+}
+
+type PortBuilderOption func(portBuilder *corev1.ServicePort)
+
+func WithTargetPort(targetPort int32) PortBuilderOption {
+	return func(servicePort *corev1.ServicePort) {
+		servicePort.TargetPort = intstr.FromInt32(targetPort)
+	}
+}
+
+func WithAppProtocol(proto *string) PortBuilderOption {
+	return func(servicePort *corev1.ServicePort) {
+		servicePort.AppProtocol = proto
+	}
+}
+
+func WithProtocol(proto corev1.Protocol) PortBuilderOption {
+	return func(servicePort *corev1.ServicePort) {
+		servicePort.Protocol = proto
+	}
+}
+
+func ComponentType(name string) string {
+	// components have a name like:
+	// - mycomponent/custom
+	// - mycomponent
+	// we extract the "mycomponent" part and see if we have a parser for the component
+	if strings.Contains(name, "/") {
+		return name[:strings.Index(name, "/")]
+	}
+
+	return name
+}
+
+func PortFromEndpoint(endpoint string) (int32, error) {
+	var err error
+	var port int64
+
+	r := regexp.MustCompile(":[0-9]+")
+
+	if r.MatchString(endpoint) {
+		port, err = strconv.ParseInt(strings.Replace(r.FindString(endpoint), ":", "", -1), 10, 32)
+
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	if port == 0 {
+		return 0, PortNotFoundErr
+	}
+
+	return int32(port), err
+}
+
+type ComponentPortParser interface {
+	// Ports returns the service ports parsed based on the exporter's configuration
+	Ports(logger logr.Logger, config interface{}) ([]corev1.ServicePort, error)
+
+	// ParserType returns the name of this parser
+	ParserType() string
+
+	// ParserName is an internal name for the parser
+	ParserName() string
+}
+
+func LoadMap[T any](m interface{}, in T) error {
```

Comment:
what would the direct conversion look like? do you mean doing it when we parse the config?

Reply:
What this function is doing, in essence, is turning a map into a given hierarchical struct. It would be cleaner to either use similar generic conversion logic (which is a bunch of reflection internally) without the serialization step. [Mapstructure](https://github.com/mitchellh/mapstructure) instead of json, in essence.